### PR TITLE
Fix GC hole with multi-reg local var stores

### LINKED ISSUE #65476: Test failure System.IO.Compression.ZLibStreamUnitTests.ReadWrite_Success
Run: [runtime-coreclr libraries-jitstress 20220215.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1615617&view=ms.vss-test-web.build-test-results-tab&runId=44957626&paneView=debug&resultId=133897)

Failed test:
```
net7.0-Linux-Release-arm64-CoreCLR_checked-zapdisable-(Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-20210531091519-97d8652

- System.IO.Compression.ZLibStreamUnitTests.ReadWrite_Success(mode: AsyncAPM, writeSize: 10240, startWithFlush: False)

```
**Error message:**
```
System.NullReferenceException : Object reference not set to an instance of an object.


Stack trace
   at System.IO.ConnectedStreams.BidirectionalStreamBufferStream.Dispose(Boolean disposing) in /_/src/libraries/Common/tests/System/IO/ConnectedStreams.cs:line 299
   at System.IO.Stream.Close()
   at System.IO.Compression.DeflateStream.Dispose(Boolean disposing) in /_/src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:line 655
   at System.IO.Stream.Close()
   at System.IO.Compression.ZLibStream.Dispose(Boolean disposing) in /_/src/libraries/System.IO.Compression/src/System/IO/Compression/ZLibStream.cs:line 252
   at System.IO.Stream.Close()
   at System.IO.Tests.ConnectedStreamConformanceTests.ReadWrite_Success(ReadWriteMode mode, Int32 writeSize, Boolean startWithFlush) in /_/src/libraries/Common/tests/StreamConformanceTests/System/IO/StreamConformanceTests.cs:line 1729
--- End of stack trace from previous location ---
```

### PR DESCRIPTION
**Pull request by BruceForstall**
Change #64857 exposed an existing problem where when generating code
for a multi-reg GT_STORE_LCL_VAR, if the first register slot was not
enregistered, but the second or subsequent slots was, and those non-first
slots contained GC pointers, we wouldn't properly add those GC pointers
to the GC tracking sets. This led to cases where the register lifetimes
would be killed in the GC info before the actual lifetime was complete.

The primary fix is to make `gtHasReg()` handle the `IsMultiRegLclVar()`
case. As a side-effect, this fixes some LSRA dumps that weren't displaying
multiple registers properly.

There are about 50 SPMI asm diffs on win-arm64 where register lifetimes
get extended, fixing GC holes.

I also made `GetMultiRegCount()` handle the `IsMultiRegLclVar()` case.

I made a number of cleanup changes along the way:
1. Fixed two cases of calling `gcInfo.gcMarkRegSetNpt` with regNumber, not regMaskTP
2. Marked some functions `const`
3. Improved some comments
4. Changed "ith" to "i'th" in comments which still doesn't read great,
but at least I'm not left trying to parse "ith" as an English word.
5. Use `OperIsScalarLocal()` more broadly
6. Renamed `gtDispRegCount` to `gtDispMultiRegCount` to make it clear
it only applies to the multi-reg case.

Fixes #65476.

### CODE REVIEW

**Comment by kunalspathak** [comment]
I will take a look on Monday.

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -655,6 +657,20 @@ bool GenTree::gtHasReg() const
             }
         }
     }
+    else if (IsMultiRegLclVar())
+    {
+        assert(comp != nullptr);
+        const GenTreeLclVar* lclNode  = AsLclVar();
+        const unsigned       regCount = GetMultiRegCount(comp);
+        for (unsigned i = 0; i < regCount; i++)
```
Comment (kunalspathak) [inline src/coreclr/jit/gentree.cpp:665]: For consistency, add a comment

```suggestion
        // A Multi-reg local vars is said to have regs,
        // if it has valid regs in any of the positions.
        for (unsigned i = 0; i < regCount; i++)
```

**Review by kunalspathak — COMMENTED** [review]
Added a minor nit. I have kicked of superpmi replay to make sure we at least run the jitstressregs.

> There are about 50 SPMI asm diffs on win-arm64 where register lifetimes
get extended, fixing GC holes.

Yes, so there are no code size or perfscore diffs, but just the register lifetimes differences? Wonder why asmdiffs (coredistools) is even thinking that there is code difference?

**Review by kunalspathak — APPROVED** [review]
superpmi-replay is clean.

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -602,23 +602,25 @@ void GenTree::CopyReg(GenTree* from)
...
+//    comp - Compiler instance. Required for multi-reg lcl var; ignored otherwise.
 //
 // Return Value:
 //    Returns true if the node was assigned a register.
 //
-//    In case of multi-reg call nodes, it is considered
-//    having a reg if regs are allocated for all its
+//    In case of multi-reg call nodes, it is considered having a reg if regs are allocated for ALL its
 //    return values.
+//    REVIEW: why is this ALL and the other cases are ANY? Explain.
```
Comment (echesakovMSFT) [inline src/coreclr/jit/gentree.cpp:615]: Do you intend to keep this comment?

Comment (BruceForstall) [reply src/coreclr/jit/gentree.cpp:615]: Yes, because I don't know the answer :-)

Comment (echesakovMSFT) [reply src/coreclr/jit/gentree.cpp:615]: Ah, okay

**Pull request merged**